File: Behavior/Visitor/Cpp/visitor.cpp
#include <iostream>
#include <vector>

// 元素接口
class Visitor; // 前置声明

class Element
{
public:
    virtual ~Element() = default;
    virtual void Accept(Visitor &visitor) = 0;
};

// 角色
class Character : public Element
{
public:
    void Accept(Visitor &visitor) override;

    void InteractWithCharacter()
    {
        std::cout << "Character interaction." << std::endl;
    }
};

// 敌人
class Enemy : public Element
{
public:
    void Accept(Visitor &visitor) override;

    void InteractWithEnemy()
    {
        std::cout << "Enemy interaction." << std::endl;
    }
};

// 道具
class Item : public Element
{
public:
    void Accept(Visitor &visitor) override;

    void InteractWithItem()
    {
        std::cout << "Item interaction." << std::endl;
    }
};

// 访问者接口
class Visitor
{
public:
    virtual ~Visitor() = default;
    virtual void VisitCharacter(Character &character) = 0;
    virtual void VisitEnemy(Enemy &enemy) = 0;
    virtual void VisitItem(Item &item) = 0;
};

// 交互访问者
class InteractionVisitor : public Visitor
{
public:
    void VisitCharacter(Character &character) override
    {
        character.InteractWithCharacter();
    }

    void VisitEnemy(Enemy &enemy) override
    {
        enemy.InteractWithEnemy();
    }

    void VisitItem(Item &item) override
    {
        item.InteractWithItem();
    }
};

void Character::Accept(Visitor &visitor)
{
    visitor.VisitCharacter(*this);
}

void Enemy::Accept(Visitor &visitor)
{
    visitor.VisitEnemy(*this);
}

void Item::Accept(Visitor &visitor)
{
    visitor.VisitItem(*this);
}

int main()
{
    std::vector<Element *> elements;
    elements.push_back(new Character());
    elements.push_back(new Enemy());
    elements.push_back(new Item());

    InteractionVisitor visitor;

    for (auto &element : elements)
    {
        element->Accept(visitor);
    }

    // 清理
    for (auto &element : elements)
    {
        delete element;
    }

    return 0;
}

File: Behavior/Visitor/Lua/visitor.lua
-- 元素接口
Element = {}
function Element:new()
    local obj = {}
    setmetatable(obj, self)
    self.__index = self
    return obj
end
function Element:accept(visitor)
    -- 将由具体元素覆盖
end

-- 角色
Character = Element:new()
function Character:accept(visitor)
    visitor:visitCharacter(self)
end

-- 敌人
Enemy = Element:new()
function Enemy:accept(visitor)
    visitor:visitEnemy(self)
end

-- 道具
Item = Element:new()
function Item:accept(visitor)
    visitor:visitItem(self)
end

-- 访问者接口
Visitor = {}
function Visitor:new()
    local obj = {}
    setmetatable(obj, self)
    self.__index = self
    return obj
end
function Visitor:visitCharacter(character)
    -- 将由具体访问者覆盖
end
function Visitor:visitEnemy(enemy)
    -- 将由具体访问者覆盖
end
function Visitor:visitItem(item)
    -- 将由具体访问者覆盖
end

-- 游戏规则访问者
GameRulesVisitor = Visitor:new()
function GameRulesVisitor:visitCharacter(character)
    print("Applying game rules to a character")
    -- 这里添加实际的游戏规则逻辑
end
function GameRulesVisitor:visitEnemy(enemy)
    print("Applying game rules to an enemy")
    -- 这里添加实际的游戏规则逻辑
end
function GameRulesVisitor:visitItem(item)
    print("Applying game rules to an item")
    -- 这里添加实际的游戏规则逻辑
end


function main()
    local elements = {Character:new(), Enemy:new(), Item:new()}

    local gameRulesVisitor = GameRulesVisitor:new()

    for _, element in ipairs(elements) do
        element:accept(gameRulesVisitor)
    end
end

main()
